lib/src/signed_video_h26x_auth.c

```diff
@@ -167,56 +158,47 @@ prepare_for_link_and_gop_hash_verification(signed_video_t *self, h26x_nalu_list_
 
   // Start with the first item in the NALU list.
   item = nalu_list->first_item;
+  svrc_t status = SV_UNKNOWN_FAILURE;
+  SV_TRY()
+    // At the start of the GOP, initialize the |crypto_handle| to enable
+    // sequentially updating the hash with more NALUs.
+    SV_THROW(openssl_init_hash(self->crypto_handle, true));
+    // Iterate through the NALU list until the end of the current GOP or SEI item is found.
+    while (item && !(found_next_gop || found_item_after_sei)) {
+      // Skip non-pending items
+      if (item->validation_status != 'P') {
+        item = item->next;
+        continue;
+      }
 
-  // Iterate through the NALU list until the end of the current GOP or SEI item is found.
-  while (item && !(found_next_gop || found_item_after_sei)) {
-    // Skip non-pending items
-    if (item->validation_status != 'P') {
-      item = item->next;
-      continue;
-    }
+      // Ensure that only non-missing NALUs (which have non-null pointers) are processed.
+      assert(item->nalu);
 
-    // Ensure that only non-missing NALUs (which have non-null pointers) are processed.
-    assert(item->nalu);
+      // Track when the current item follows the SEI or marks the beginning of the next GOP.
+      found_item_after_sei = (item->prev == sei);
+      found_next_gop = (item->nalu->is_first_nalu_in_gop && !item->need_second_verification);
 
-    // Track when the current item follows the SEI or marks the beginning of the next GOP.
-    found_item_after_sei = (item->prev == sei);
-    found_next_gop = (item->nalu->is_first_nalu_in_gop && !item->need_second_verification);
+      // Skip GOP SEI items as they do not contribute to the GOP hash.
+      if (item->nalu->is_gop_sei) {
+        item = item->next;
+        continue;
+      }
+      hash_to_add = item->need_second_verification ? item->second_hash : item->hash;
+      // Since the GOP hash is initialized, it can be updated with each incoming NALU hash.
+      SV_THROW(openssl_update_hash(self->crypto_handle, hash_to_add, hash_size, true));
+      item->used_in_gop_hash = true;  // Mark the item as used in the GOP hash
 
-    // Skip GOP SEI items as they do not contribute to the GOP hash.
-    if (item->nalu->is_gop_sei) {
       item = item->next;
-      continue;
     }
+    // Finally, mark the SEI item as used in the GOP hash
```

Comment:
Why can't you also finalize the GOP hash while you're at it? You have both init and update in this function. To me it would make sense to also have finalize here as well.